Add handshake fuzz targets

jfoote opened the pull request:
This PR adds two new fuzz targets based on work from @mimoo (thanks!). Here is a quick summary of code changes:

- Adds two new fuzz targets
- Adds corresponding build logic for `-DBUILD_FUZZERS` 
- Adds seed corpora for the fuzz targets
- Makes some minor patches to `lib/picotls.c` (and some accompanying files) to short-circuit some logic for testing in fuzzing builds
- Adds a simple fuzzing `README.md`

Thanks to @deweerdt for initial feedback.

I am planning to make a follow up PR to [h2o](https://github.com/h2o/h2o) and then [oss-fuzz](https://github.com/google/oss-fuzz) to get these running continuously.

deweerdt commented on CMakeLists.txt:
```diff
@@ -76,4 +76,20 @@ IF (BUILD_FUZZER)
+
+    SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -D_GNU_SOURCE")
+    INCLUDE_DIRECTORIES(t include ${OPENSSL_INCLUDE_DIR})
+    ADD_DEFINITIONS(-DPTLS_FUZZ_HANDSHAKE)
```
I believe this needs to be `-DPTLS_FUZZ_HANDSHAKE=1`, doesn't it?

jfoote replied:
It seemed to work without the `=1`, but I changed it anyway -- better to be explicit I think.

deweerdt commented on include/picotls.h:
```diff
@@ -656,6 +659,8 @@ typedef struct st_ptls_handshake_properties_t {
 #pragma warning(pop)
 #endif
 
+void ptls_fuzzing();
```
What is this function used for?

jfoote replied:
It is leftover cruft from the original implementation of the fuzz targets -- removed.

jfoote commented:
I fixed a bug with a macro that was causing CI to fail and addressed @deweerdt's comments -- thanks for the feedback! If you'd like to see any other changes here just let me know.

kazuho commented on lib/picotls.c:
```diff
@@ -525,40 +525,11 @@ int ptls_buffer_push_asn1_ubigint(ptls_buffer_t *buf, const void *bignum, size_t
+#include "aead_crypt_fuzz.c.h"
+#else
+#include "aead_crypt.c.h"
+#endif
```
Can we move `aead_encrypt` and `aead_decrypt` back to picotls.c and do `#if PTLS_FUZZ_HANDSHAKE` inside the functions?

We are having `#if`s within the source file at other places (see below), and I wonder if it's worth having different source files for this specific case.

jfoote replied:
> Can we move aead_encrypt and aead_decrypt back to picotls.c and do #if PTLS_FUZZ_HANDSHAKE inside the functions?

Sure, done in https://github.com/h2o/picotls/pull/197/commits/6eeea1248b7618f5ae4d9aad917cd6e70ff354bd

> We are having #ifs within the source file at other places (see below), and I wonder if it's worth having different source files for this specific case.

Good question -- perhaps we can decide pending resolution to the other issues you have flagged.

jfoote replied:
> We are having #ifs within the source file at other places (see below), and I wonder if it's worth having different source files for this specific case.

After resolving the other comments I don't think we don't need any other `#if`s

kazuho commented on lib/picotls.c:
```diff
@@ -2752,11 +2723,13 @@ static int try_psk_handshake(ptls_t *tls, size_t *psk_index, int *accept_early_d
-            continue;
-        if (now - issue_at > (uint64_t)tls->ctx->ticket_lifetime * 1000)
-            continue;
+        if (!PTLS_FUZZ_HANDSHAKE) {
```
Would it be possible use a special `get_time` callback that returns the expected value?

jfoote replied:
I am not sure how to do this unobtrusively. It seems the special `get_time` would need to know the `issue_at` value from the fuzzed data supplied in each fuzz iteration to work properly. I thought of some options to do this, but none were great. One idea would be to call `decode_session_identifier` in `encrypt_ticket_cb_fake` and store the resulting `issue_at` value in a static global variable that could then be accessed from the special `get_time` callback, but that is a bit hacky since we are calling the parser we intend to test.

Did you have something in mind perhaps?

jfoote replied:
For posterity we discussed this via Slack; this should address the issue: https://github.com/jfoote/picotls/commit/2d065a20592158578e8e4c562d1db20bc93b86e0

kazuho commented on lib/picotls.c:
```diff
@@ -2110,7 +2081,7 @@ static int handle_certificate(ptls_t *tls, const uint8_t *src, const uint8_t *en
     });
 
-    if (num_certs != 0 && tls->ctx->verify_certificate != NULL) {
+    if (!PTLS_FUZZ_HANDSHAKE && num_certs != 0 && tls->ctx->verify_certificate != NULL) {
```
Would it be possible to set verify_certificate callback to NULL in the fuzzer?

jfoote replied:
Good idea -- done in https://github.com/h2o/picotls/pull/197/commits/ae326e627a111253893f5f7629c5bd1a43179e0d

kazuho commented on lib/picotls.c:
```diff
@@ -2237,7 +2208,7 @@ static int handle_certificate_verify(ptls_t *tls, ptls_iovec_t message, const ch
     }
     signdata_size = build_certificate_verify_signdata(signdata, tls->key_schedule, context_string);
-    if (tls->certificate_verify.cb != NULL) {
+    if (!PTLS_FUZZ_HANDSHAKE && tls->certificate_verify.cb != NULL) {
```
Do we need this change? I ask this because IIUC it is set only through the invocation of `verify_certificate.cb`. However, the change above skips the invocation.

jfoote replied:
Makes sense -- I removed this check in https://github.com/h2o/picotls/pull/197/commits/ae326e627a111253893f5f7629c5bd1a43179e0d and verified the coverage remains the same/before after via a small test run for added confidence

jfoote commented:
In analyzing how to best address @kazuho's comments I discovered the two `PTLS_FUZZ_HANDSHAKE` conditionals above never seem to be exercised by the fuzzers. I compiled the `fuzz-client-hello` fuzz target (which I believe was intended to exercise the logic) to emit clang coverage and then ran it for six hours as a test. It looks like fuzzing never progresses past the client hello version check:

```
[lib/picotls.c]
[line|hits   |]
 2666|   589M|        }
 2667|   589M|        src = end;
 2668|   589M|    });
 2669|   589M|
 2670|   589M|    /* check if client hello make sense */
 2671|   589M|    if (is_supported_version(ch->selected_version)) {
 2672|      0|        if (!(ch->compression_methods.count == 1 && ch->compression_methods.ids[0] == 0)) {
 2673|      0|            ret = PTLS_ALERT_ILLEGAL_PARAMETER;
 2674|      0|            goto Exit;
```

I have some thoughts on addressing this but it may be delayed until after the US holiday. In the meantime if anyone has any ideas please feel free to chime in. Or feel free to hack on it directly of course.

Thanks for all the help so far @kazuho and @deweerdt

mimoo commented:
Hey! This makes sure that a version is both supported by the client and the server, and then that the compression method is set to NULL. It is quite weird that the fuzzer would get stuck there as there are definitely some inputs that are valid, but it is possible that too many of the corpus inputs are not passing this check and exercising some other path. If this seems to be the case, instead of hopping that the fuzzer creates the correct inputs, you might be able to skip the check (that is, if picotls correctly parses the client hello when the `compression_methods` value is not of the expected length of 1 byte). You could also ignore this and let the fuzzer do its thing long-term :)

jfoote commented:
Thanks for chiming in @mimoo. I think the corpora were swapped in the initial tarball. I coalesced all the original files (and then some) and distilled the entire set for each fuzz target in https://github.com/h2o/picotls/pull/197/commits/d398f62edb9ecd45b5d6e91068fdfd07d8d72997. Execution is proceeding past those checks now.

mimoo commented:
Nice! You should probably merge all that corpus anyway and minimize it with the `-merge=1` option.

jfoote commented:
> Nice! You should probably merge all that corpus anyway and minimize it with the -merge=1 option.

@mimoo This is what I was referring to by corpus distillation in [my previous comment](https://github.com/h2o/picotls/pull/197#issuecomment-458208931) -- it is already done.

@kazuho @deweerdt I think this should be good to go pending any further feedback.

kazuho commented:
@jfoote Thank you for the changes! Merged to master.
